dask_expr/tests/test_distributed.py

```diff
@@ -354,3 +355,186 @@ def test_func(n):
     result = await result
     expected = pd.DataFrame({"a": [4951, 4952, 4953, 4954]})
     pd.testing.assert_frame_equal(result, expected)
+
+
+@gen_cluster(client=True)
+async def test_p2p_shuffle_reuse(c, s, a, b):
+    pdf = pd.DataFrame({"a": [1, 2, 3, 4, 5, 6] * 10, "b": 2, "e": 2})
+    df = from_pandas(pdf, npartitions=10)
+    q = df.shuffle("a")
+    q = q.fillna(100)
+    q = q.a + q.a.sum()
+    # Only one IO node since shuffle consumes
+    _check_consumer_node(q, 1)
+    _check_consumer_node(q, 2, consumer_node=P2PShuffle)
+    x = c.compute(q)
+    x = await x
+
+    expected = pdf.fillna(100)
+    expected = expected.a + expected.a.sum()
+    pd.testing.assert_series_equal(x.sort_index(), expected)
+
+    # Check that we have 1 shuffle barrier but 20 p2pshuffle tasks for the output
```

Comment:
IIUC, this PR introduces functionality that relies on being able to read shuffle outputs several times. The fact that this seems to work with disk-based P2P is a lucky implementation detail but not guaranteed to work, let alone tested. Before releasing, we should test this. (It will also not work with in-memory P2P, but that's currently not supported in `dask-expr` anyway.)